hashes/src/hkdf.rs

```diff
@@ -0,0 +1,192 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! HMAC-based Extract-and-Expand Key Derivation Function (HKDF).
+//!
+//! Implementation based on RFC5869, but the interface is scoped
+//! to BIP324's requirements.
+
+use core::fmt;
+
+#[cfg(all(feature = "alloc", not(feature = "std")))]
+use alloc::vec;
+#[cfg(all(feature = "alloc", not(feature = "std")))]
+use alloc::vec::Vec;
```

Comment:
Something looks broken with the tools, I'm getting a clippy warning for this line but removing it gives a `Vec` not found error - one solution is to remove this line and use `alloc::vec::Vec` in `expand_to_len`.

Reply:
Hmm finding this one to be a bit tricky. I think it has to do with how `alloc` is re-exported in `std`. If the standard flags are being used, we want to just pull from the prelude which I believe is technically `std::vec::`. Tricky getting the tools to align on this one, but I'll keep iterating on things.

Reply:
Often the "pick one of 'redundant import' or 'missing import'" clippy behavior is caused by something that's imported outside of the test module but only used inside of the test module. Is something like that happening?

Reply:
I am a bit green to the "no_std" stuff, so sorry if I am off on this one, but this is my theory so far.

The clippy errors are only happening on the `nightly` and `beta` toolchains, it is fine on `stable` for now:

```
[../rust-bitcoin add-hdkf]$ cargo +beta clippy --package bitcoin_hashes --all-targets --no-default-features --features alloc
warning: /home/njohnson/grd/rust-bitcoin/bitcoin/Cargo.toml: version requirement `0.105.0+25.1` for dependency `bitcoinconsensus` includes semver metadata which will be ignored, removing the metadata is recommended to avoid confusion
warning: the item `Vec` is imported redundantly
  --> hashes/src/hkdf.rs:13:5
   |
13 | use alloc::vec::Vec;
   |     ^^^^^^^^^^^^^^^
  --> /rustc/c1195518c8ad7f37d15ecb82f39dec3da97f9c2f/library/std/src/prelude/mod.rs:125:13
   |
   = note: the item `Vec` is already defined here
   |
   = note: `#[warn(unused_imports)]` on by default

warning: `bitcoin_hashes` (lib test) generated 1 warning
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.05s
```

From the errors, it looks like the "alloc" imports are conflicting with the std prelude. This is surprising for me since it appears like the imports should only be enabled if there is no std predule included: `#[cfg(all(feature = "alloc", not(feature = "std")))]`. I am not sure how the std predule is getting there tbh, given the `--no-default-features --features alloc` flags.

I found this issue though which appears to be effecting a lot of people on the nightly and beta toolchains: https://github.com/rust-lang/rust/issues/121362. From there it links to an underlying issue: https://github.com/rust-lang/rust/issues/121708 (which it looks like ya posted on before @apoelstra for a different reason). 

Fwiw, I have tried swapping out the no_std pattern in `hashes` like it suggests in these threads, but I haven't had any luck fixing the clippy issue:

```
#![no_std]

#[cfg(feature = "std")]
extern crate std;
```

So tldr; it appears the std prelude is somehow getting added (or is counted as added but actually isn't?) in the beta and nightly toolchains, but not clear to me how.

Reply:
Yes, nightly/beta have a ton more "redundant import" lints, in particular around things that are feature-gated.

Without looking more closely, to resolve this you should grep the codebase for `use*Vec` and find other places where we import `Vec` without warnings, and copy whatever feature-gates are on those uses.

Reply:
I think I got to the bottom of things. With your guy's pointers, I saw that the issues were only occurring for the `tests` target. The crux is that the root module hashes/src/lib.rs treats the test target a little different for disabling no_std: `#![cfg_attr(all(not(test), not(feature = "std")), no_std)]`. This doesn't play nice with the import in hashes/src/hkdf.rs: `#[cfg(all(feature = "alloc", not(feature = "std")))]`. If we run clippy with only the alloc flag for the tests target, both the std predule and the alloc imports are triggered.

I went with the simplest solution in a17ab8c934d6c86adbd3f4d0067c867bb40e26dc to resolve the redundant import. I noticed in the `io` package, which is very similar, the `tests` module is using a wildcard `super::*` to import the functions to test. I didn't realize there is a semantic difference between using a wildcard and explicitly importing (e.g. `crate::hkdf::Hkdf`), I haven't found the best docs on it yet, but I believe the wildcard has some internal logic to *not* import something if it causes a namespace issue. So it protects us in this case where it appears an explicit import does not.

The wildcard protection wasn't immediately clear to me, so I went on a bit of journey I'll document here, but feel free to skip! 

Another way to fix this issue is to extend the alloc guards with a test clause: `#[cfg(all(feature = "alloc", not(feature = "std"), not(tests))]`. I didn't see this pattern in rust-bitcoin though, so stuck with the wildcards for now.

One difference I did notice between the `io` package and `hashes` is how the extern crate alloc is imported. In `io`, it only checks for the alloc flag: `#[cfg(feature = "alloc")]`. In `hashes`, it also checks that there isn't a std flag: `#[cfg(all(feature = "alloc", not(feature = "std")))]`. I ran some tests and am still not sure what the extra check is protecting us from. But one thing I did notice is that if that part is dropped, it makes it easier to know when the alloc crate is in the namespace. For example, the `expand_to_len` function in this PR could assume the alloc crate is available and be simplified to the following, no longer needing the more complex import statements attempting to tease apart when the std prelude is there or not:

```
#[cfg(feature = "alloc")]
    pub fn expand_to_len(&self, info: &[u8], len: usize) -> Result<alloc::vec::Vec<u8>, MaxLengthError> {
        let mut okm = alloc::vec![0u8; len];
        self.expand(info, &mut okm)?;
        Ok(okm)
    }
```

Reply:
Good digging. Some other things you might find interesting/confusing:
- tests always have `std` crate available

Putting this test in `hashes/src/lib.rs` works even with `cargo test --no-default-features`

```rust
    #[test]
    fn prove_std_available() {
        #[cfg(feature = "std")]
        println!("std is enabled");

        #[cfg(feature = "alloc")]
        println!("alloc is enabled");

        let _ = std::string::String::new();
    }
```

- test cade is checked twice, once with `test` enabled and once without (I'm a bit hazy on the details

Can see by using this test
```rust
     // If test code was never built with `test` not enabled then this code should build just fine.
    #[test]
    fn panic() {
        #[cfg(all(feature = "std", feature = "alloc", not(test)))]
        {
            this wont build
        }
        println!("this builds just fine")
    }
```

Calling `cargo test  --no-default-features --features=alloc,std`

Reply:
Wow, that second point of yours is really surprising...wonder what is going on there.

Reply:
> Putting this test in `hashes/src/lib.rs` works even with `cargo test --no-default-features`

Yes, the test works, but it doesn't print either message.

> * test code seems to be checked twice, once with `test` enabled and once without

The *code* is checked twice but the tests are not run twice.